lib/view_component/base.rb

```diff
@@ -293,6 +351,65 @@ def with_content_areas(*areas)
         self.content_areas = areas
       end
 
+      # support initalizing slots as:
+      #
+      # with_slot(
+      #   :header,
+      #   collection: true|false,
+      #   class_name: "Header" # class name string, used to instantiate Slot
+      # )
+      def with_slot(*slot_names, collection: false, class_name: nil)
```

Comment:
The `class_name` argument here actually confused me for a good minute or two since I was in the mindset of "this is for the view" and thought it was like HTML `class="foo bar"` and not a Ruby class name.

This almost makes me think that `with_slot` might be more clear as a DSL instead of a method that associates with another class?

eg:

```ruby
with_slot :header, collection: false do |arg1, arg2, arg3|
end
```

I think that may also have the added benefit of removing nested classes in components but would also keep all slot logic in a single place in the component.

That's probably a huge change and has its own challenges, but I thought it was worth bringing up the confusion I ran into here.

Reply:
@BlakeWilliams I share your confusion. The `class_name` API is meant to mimic ActiveRecord: https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-class-name. 

As you mentioned, there has been a lot of debates about the merits of a DSL approach vs. this Class-based approach. I could honestly go either way at this point, but I keep coming back to the fact that Slots are, in my mind, not meant to stand on their own conceptually, let alone be referenced/exist outside the scope of a single component.

@bbugh you seemed to have the strongest feelings about having the Class-based implementation. What's your feeling at this point?

Reply:
> not meant to stand on their own conceptually, let alone be referenced/exist outside the scope of a single component.

That makes sense to me, and I think builds on your above comment about slots just having a simple `content` accessor.

I wonder if both of those points help make a stronger case for the DSL API over Ruby objects?

Personally I think the DSL style helps reinforce "this is not a component, it's just a smaller, isolated part of a component with its own encapsulated behavior and responsibilities".

Reply:
I agree that DSLs are nice, but I think it is very important not to use a DSL here and allow `class_name` to be set. I explained my reasoning here, let me know if you have more questions. https://github.com/github/view_component/pull/348#issuecomment-643839329

In short, using the DSL in this way means that the API must choose to either allow component subclasses to replace a slot OR to extend a slot, but it cannot support both. With regular Ruby classes, it doesn't limit the usage and any future use cases for third party libraries.

Reply:
@bbugh Did you see my comment here? https://github.com/github/view_component/pull/348#issuecomment-643862606 I tried it out and I'm pretty sure it wont work as you want it to regardless :/

Reply:
> The class_name argument here actually confused me for a good minute or two since I was in the mindset of "this is for the view" and thought it was like HTML class="foo bar" and not a Ruby class name.

Heh, you're right, good feedback..

Reply:
> In short, using the DSL in this way means that the API must choose to either allow component subclasses to replace a slot OR to extend a slot, but it cannot support both. With regular Ruby classes, it doesn't limit the usage and any future use cases for third party libraries.

That's a really good point, but I wonder if replacing slots wouldn't be good enough in most cases? I'm also curious how slots would be used and if they'll end up being complex enough that extending will be a common need/want.

If we did go down the DSL path, do we think it'd be able to address the extending problem if it became an issue by adding more DSL methods? e.g. `extend_slot :header`.  (I will admit, the `extend_slot` idea does get a bit into "reinventing the wheel" territory inheritance wise 😄)

I agree that we could go either way API wise, but I do feel like the DSL makes it clear that slots are special and is a bit more fun/clear to write in my opinion. I guess I'm not sure if it's worth trading API/UX for flexibility in this case, personally.

Reply:
@jensljungblad I'm not sure about the code not working, sorry! It sounds like maybe that's a limitation with the current setup for ViewComponent? I don't believe it's a Ruby limitation, this is a common pattern with Pundit's `Scope` nested classes. Here's code showing the different variations of what I mean:

```ruby
class Component < ViewComponent::Base
  class Subslot < ViewComponent::Slot
    def some_method
      "red"
    end
  end
end

class Child1 < Component
  # Subclass with no override
  # for demonstration purposes
  class Subslot < Subslot
    # some_method will be "red"
  end
end

class Child2 < Component
  # Subclass with override
  class Subslot < Subslot
    def some_method
      "blue"
    end
  end
end

class Child3 < Component
  # Use a different class altogether
  with_slot :subslot, class_name: 'SomeOtherSubslot'

  class SomeOtherSubslot < ViewComponent::Slot
    def some_method
      "fifty"
    end
  end
end
```

> @BlakeWilliams: That's a really good point, but I wonder if replacing slots wouldn't be good enough in most cases? I'm also curious how slots would be used and if they'll end up being complex enough that extending will be a common need/want.

Good question! To clarify, I'm not arguing this for the sake of complexity, if a slot gets complex, they can use a component! I'm advocating for `class_name` in order to be able to use standard Ruby code in components when you need to write standard Ruby code in subclasses (like the above example code).

For one example, imagine if somebody released a form library like `simple_form` made with `view_component`. In order to adopt it, you'd need to know that you could extend/replace parts as-needed. You might need to replace a slot, or you might need to extend a slot, both of which are easy to do if it's just Ruby, but with the DSL idea, `view_component` can only support one or the other.

For another example, let's say you need some of your custom `Slot`s in your component tree to descend from an `InstrumentedSlot`, because you want to instrument certain calls to `content` (like for caching). That's easy with plain Ruby, but impossible with the DSL idea.

One more example, let's say you're making a component that has multiple different kinds of slots, but they mostly share the same behavior, and you want to have a parent class for all of those slot types. That's easy with Ruby, but impossible with the DSL idea.

I made these up off the top of my head and could come up with more, but the point I'm wanting to make is that we can't predict how people might want to use it, so we shouldn't unnecessarily limit it.

I am all for DSLs, I think they're great! I think they're one of the best features of Ruby's culture. I also think it would be a mistake to do here, as I hope the above examples demonstrate. As far as I can tell, the `with_slot` class-block API doesn't give us anything except that it (arguably) looks a little nicer, but it takes away a lot of choice and power from the users. I think the best thing to do is to let people write `class Thing < ViewComponent::Slot`, specify the slot classes, put their code where they want, and be able to just write Ruby to get 'er done. ¢¢

Reply:
> I'm advocating for class_name in order to be able to use standard Ruby code in components when you need to write standard Ruby code in subclasses (like the above example code).

I see. Would the proposed `extend_slot` API idea also solve that problem though assuming it comes up as one once folks are actively using the API?

> For one example, imagine if somebody released a form library like simple_form made with view_component. In order to adopt it, you'd need to know that you could extend/replace parts as-needed. You might need to replace a slot, or you might need to extend a slot, both of which are easy to do if it's just Ruby, but with the DSL idea, view_component can only support one or the other.

I'm trying to imagine what a simple form library using slots would look like but feel like I'm struggling a bit. When I think of slots, I think slots are small parts of components that you need to change that are separate from the main content and may require a small bit of logic/arguments, but not anything too complex. Do you have any rough ideas what a flexible simple form slot library would look like at least in terms of usage?

> For another example, let's say you need some of your custom Slots in your component tree to descend from an InstrumentedSlot, because you want to instrument certain calls to content (like for caching). That's easy with plain Ruby, but impossible with the DSL idea.

> One more example, let's say you're making a component that has multiple different kinds of slots, but they mostly share the same behavior, and you want to have a parent class for all of those slot types. That's easy with Ruby, but impossible with the DSL idea.

I think those are really good, and valid points, but it does make me ask if class based slots are the correct abstraction for those problems.

> I am all for DSLs, I think they're great! I think they're one of the best features of Ruby's culture. I also think it would be a mistake to do here, as I hope the above examples demonstrate. As far as I can tell, the with_slot class-block API doesn't give us anything except that it (arguably) looks a little nicer, but it takes away a lot of choice and power from the users. I think the best thing to do is to let people write class Thing < ViewComponent::Slot, specify the slot classes, put their code where they want, and be able to just write Ruby to get 'er done. ¢¢

I've been thinking about slots a good bit in my free time and I'm honestly having a hard time differentiating 
slots from components with the class based API. We're trying to get extensibility and complexity out of them which components already give us. The primary difference seems to be that they lack a template and convention defines them *inside* of the component instead of in a separate file? (that does give me some file length and manageability concerns though)

Breaking that down, I guess my questions would be:

* When slots are class based, is there a benefit to using them over components? If so, are the trade-offs of having another concept, longer/more complex component files, etc. worth it?
* If we did go down the DSL path, would some of this be solved by being able to pass component instances to `component.slot` calls in the template?
* Would slots being simple/limited be a feature, and would it help differentiate them from components and when to use one over the other?

For context, I have written a lot of React so that may be influencing my thought process here a good bit. 😅 

One final thought on all of this, I wonder if it would make sense for these to be includes, or separate libraries you have to opt-in to before they're part of the stable API? It's really difficult to talk about these concepts concretely without having used them and experience the pros and cons with real world usage.

Reply:
> It's really difficult to talk about these concepts concretely without having used them and experience the pros and cons with real world usage.

This is why I've indicated that the Slots APIs are experimental in the README. They can and should change before the next major release (`v3.0.0`).

Reply:
> The primary difference seems to be that they lack a template and convention defines them inside of the component instead of in a separate file?

That's how I feel, too.

Reply:
@BlakeWilliams what do you think about supporting _both_ APIs? Perhaps we could have the DSL be the default, recommended path, but we could also support `class_names` for more complex use cases?

Reply:
@bbugh In your last example you are inheriting though, not re-opening the class! Are you saying then that you expect the `with_slot` declaration in the parent to resolve to the subclassed slot class? Is that intuitive?

Reply:
Could an experimental API just support both for now, then be tweaked once there are some actual problems/use-cases?

```ruby
with_slot :header

with_slot :body do
  def initialize(param:); end
end

with_slot :footer, class_name: "Footer" 
```

Reply:
> This is why I've indicated that the Slots APIs are experimental in the README. They can and should change before the next major release (`v3.0.0`).

I wonder if it would make sense to make it opt-in if it's experimental? I think that would make it more explicit that it will likely change and isn't part of the stable API? It would also let us version the API, so we could do something like `include ViewComponent::Slots::V2` which would allow us to support older version for a short time until we get to a stable API.

It's also possible I'm overthinking it. 😄 

> @BlakeWilliams what do you think about supporting _both_ APIs? Perhaps we could have the DSL be the default, recommended path, but we could also support `class_names` for more complex use cases?

I think that's a good idea! I do wonder if it might be too easy of a fallback initially and could push us towards using `class_name` instead of finding work arounds or other possible solutions DSL wise, but it still sounds like a solid plan. 👍

Reply:
If you do the API as an option at all it's going to cause these problems. 😌 That leaves it up to the component author and if they pick the save-one-line-DSL then end users can't decide for themselves how to use it. 

I'm sorry I'm not explaining this in a way that makes sense, the issues seem so obvious to me and I'm clearly failing to convey it well. I'll shut up about it, and hope that you guys encounter the problems yourselves before you launch it and it can't be undone.

Reply:
@bbugh I understand what you mean, and that it would be up to the component author to make it fully extensible. I just have a hard time thinking about this without actually implementing/trying it out.

I'd be okay to merge without the DSL first, just to have a starting point.

Reply:
> If you do the API as an option at all it's going to cause these problems. 😌 That leaves it up to the component author and if they pick the save-one-line-DSL then end users can't decide for themselves how to use it.

Let me know if I'm understanding or not, but is the issue that the API provider (who writes the component) using the DSL API means that consumers of that API can't extend slots?

> I'm sorry I'm not explaining this in a way that makes sense, the issues seem so obvious to me and I'm clearly failing to convey it well. I'll shut up about it, and hope that you guys encounter the problems yourselves before you launch it and it can't be undone.

This is a pretty complex issue that's not exactly concrete so talking about it can difficult (I know I've struggled a bit doing so 😄) but I super appreciate your raising concerns and discussing them here and hope you continue to do so if you'd like.

I think I understand the trade-offs, and what flexibility the DSL would lose assuming my first reply in this comment is correct (if not, I'd love to try and understand it). It could be just me, but slots and components seem to have so much overlap that a class based approach makes me think that it would be easier/clearer/more flexible if slots were passed components instead of being a separate/similar concept. To me, the DSL helps make it clear that slots aren't components and really makes the responsibility of a slot, and when to use a slot more clear. At least that's how I feel after reading, discussing, and thinking about it with my very React influenced point of view.

Reply:
> I'd be okay to merge without the DSL first, just to have a starting point.

Ultimately I think I ended up leaning towards a DSL being the way to go, but I agree that merging without the DSL as a starting point to get a feel for slots would also be fine and would allow us to have some more concrete conversations with specific, and possibly real-world examples. 👍

Reply:
> @bbugh: For one example, imagine if somebody released a form library like simple_form made with view_component. In order to adopt it, you'd need to know that you could extend/replace parts as-needed. You might need to replace a slot, or you might need to extend a slot, both of which are easy to do if it's just Ruby, but with the DSL idea, view_component can only support one or the other.

I've also done a bit of React in the past. In React there are plenty of third-party component libraries (bootstrap etc) that can be used by app developers. They typically solve these things with composition instead of inheritance: https://reactjs.org/docs/composition-vs-inheritance.html. I'd be interested in experimenting with some real world cases!

> @bbugh: For another example, let's say you need some of your custom Slots in your component tree to descend from an InstrumentedSlot, because you want to instrument certain calls to content (like for caching). That's easy with plain Ruby, but impossible with the DSL idea.

That should be possible still:

```ruby
class MyComponent
  with_slot :header do
    include(InstrumentedSlot)
  end
end

module InstrumentedSlot
  def content
    instrument { super }
  end
end
```

> @bbugh: One more example, let's say you're making a component that has multiple different kinds of slots, but they mostly share the same behavior, and you want to have a parent class for all of those slot types. That's easy with Ruby, but impossible with the DSL idea.

Should be possible with a module:

```ruby
class MyComponent
  with_slot :header do
    include(CommonBehaviour)
  end

  with_slot :body do
    include(CommonBehaviour)
  end
end
```

That being said, I still it'd be fine to start out with the class based approach and try it out!

Reply:
Maybe I'd also vote for `with_slot :header, -> { Header }` just to avoid confusion reg. Ruby class name and HTML class names.

Reply:
I also prefer the proc syntax to using a string of the name of the class in the `class_name: 'Header'` syntax. With a proc, things like autocompletion in your editor will work whereas they likely won't if you're typing the class name in a string literal. Though I don't have experience with it, I would imagine the proc form also plays nicer with Sorbet/type checking than if it's all just strings.

Reply:
@rmacklin I can see how a proc would allow for simpler autocompletion, but I worry about going that route for a couple of reasons:

1) What if the proc ends up being request-dependent? Allowing an arbitrary block of code opens us up to it executing _whatever_ a developer wants, potentially making it difficult to precompile Slots.

2) Procs are known for being slow. They are far from "free".

3) Would it really be that bad to just reference the Slot class name directly, without a proc? (like `with_slot :header, Header`) If we're loading a component, I think it's safe to assume that we want to load all of its slots. What do you think @jensljungblad @bbugh?

Reply:
> What if the proc ends up being request-dependent? Allowing an arbitrary block of code opens us up to it executing _whatever_ a developer wants, potentially making it difficult to precompile Slots.

I think we can/should prevent this. I had been thinking if we had the proc syntax, we'd call the procs to resolve the associated classes within the `compile` method. That'd ensure they're only executed once, which would prevent them from being request-dependent and would address the concern about them being slow.

Also I think the documentation could describe that the purpose of the proc is only to lazy-load the constant; if someone tries do so something crazy in the proc that would not supported and they're on their own. (FWIW, they can already execute whatever they want by overriding `const_get`, which I think is something we would not want to support.)

> Would it really be that bad to just reference the Slot class name directly, without a proc? (like `with_slot :header, Header`) If we're loading a component, I think it's safe to assume that we want to load all of its slots.

I think that forces `class Header` to be defined before you call `with_slot :header, Header`, which was considered undesirable in https://github.com/github/view_component/pull/348#issuecomment-633692353. If we're okay with that restriction, that syntax would also work well regarding autocompletion, type-checking, etc.

Reply:
Like @rmacklin said, that was my only concern with `with_slot :header, Header`, the fact that the class has to be defined at the top of the file. It's not the end of the world.

I didn't look very closely at the compile method but yes, my hope was that the resolved class could be cached somehow.

Reply:
My sense is it's powerful to have both options. I'm I'm the end user I'm likely to want to use the macros (I think of these as syntactic sugar/macros rather than a DSL). However, if I'm a Library writer I should use the API version and expose my slot classes IFF I want them to be extendable.

Reply:
I also have a pretty allergic reaction to using a Proc for the reasons @joelhawksley mentioned. I'm also concerned its not very "Rails" like I can't think of examples of this pattern. However, using strings for the names of classes is very common in AR and other places. It's not so convenient for autocompletion etc. but I'd take the safety and simplicity of the string over that.

Reply:
@jensljungblad the examples that you gave are not solutions to the problem I'm trying to explain. I will try at some point with real code to demonstrate.